### PR TITLE
fix(npm): Fix replace strategy edge-case for carets

### LINKED ISSUE #4762: npm and cargo managers "downgrade" caret dependencies in some cases
**Describe the bug**

In some cases npm or cargo managers' `getNewValue` function in `lib/versioning` "downgrades" caret ranges. "Downgraded" caret ranges still include both `fromVersion` and `toVersion` but the result itself is less than `fromVersion`.

For example if `fromVersion = 0.9.21` and `toVersion = 0.9.22` then a range `^0.9.21` will be updated to `^0.9.0`.

**Did you see anything helpful in debug logs?**

There is an example of this problem in #4620.

**To Reproduce**
Run in  `versioning/npm.spec.ts`
```javascript
import { api as npm } from '../../lib/versioning/npm';
const npmResult = npm.getNewValue('^0.9.21', 'replace', '0.9.21', '0.9.22');
console.log(npmResult);
// npmResult == '^0.9.0';
```
Run in `versioning/cargo.spec.ts`
```javascript
import { api as cargo } from './lib/versioning/cargo';
const cargoResult = cargo.getNewValue('0.9.21', 'replace', '0.9.21', '0.9.22');
console.log(cargoResult);
// cargoResult == '0.9.0';
```

**Additional context**

See [this comment](https://github.com/renovatebot/renovate/issues/4620#issuecomment-548456101).

### PR DESCRIPTION
**Pull request by renovate-testing**
## Changes:

Re-implement new value update logic. It's more verbose, but it backward-compatible with current behavior while fixing its edge-cases.

## Context:

Closes #4762

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by viceice** [comment]
@zharinov can you use your normal account for future pr's. 😏

**Comment by viceice** [comment]
🙈😅👍

File: lib/versioning/npm/index.spec.ts [hunk 1]
```diff
@@ -84,6 +84,14 @@ describe('semver.getNewValue()', () => {
       ['^1.2.3', '4.0.0', '^4.0.0'],
       ['^1.2.3', '4.5.6', '^4.0.0'],
       ['^1.0.0', '4.5.6', '^4.0.0'],
+
+      ['^0.2.3', '0.2.4', '^0.2.4'],
+      ['^2.3.0', '2.4.0', '^2.4.0'],
+      ['^2.3.4', '2.4.5', '^2.4.0'],
```
Comment (rarkins) [inline lib/versioning/npm/index.spec.ts:90]: These shouldn't really happen with `replace` though, should they?

File: lib/versioning/npm/index.spec.ts [hunk 2]
```diff
@@ -84,6 +84,14 @@ describe('semver.getNewValue()', () => {
...
+      ['^0.0.1', '0.0.2', '^0.0.2'],
+      ['^1.0.1', '2.0.2', '^2.0.0'],
+      ['^1.2.3', '1.2.3', '^1.2.3'],
+      ['^1.2.3', '1.2.0', '^1.2.3'],
```
Comment (rarkins) [inline lib/versioning/npm/index.spec.ts:94]: Not sure about this one?

Comment (zharinov) [reply lib/versioning/npm/index.spec.ts]: Well, it seems sometimes downgrades are desirable

File: lib/versioning/npm/index.spec.ts (hunk 1 above)
Comment (zharinov) [reply lib/versioning/npm/index.spec.ts]: Fixed, now it remains `^2.3.4`

**Comment by zharinov** [comment]
```typescript
expect(
  poetry.getNewValue({
    currentValue: '^1.0.0',
    rangeStrategy: 'replace',
    currentVersion: '1.0.0',
    newVersion: '1.0.7',
  })
).toEqual('^1.0.7');
```

Do we really need upgrade to `^1.0.7` here? If I understand it right, `replace` strategy should return new ranges with new versions and shouldn't just shrink old ranges.

**Comment by rarkins** [comment]
> ```ts
> expect(
>   poetry.getNewValue({
>     currentValue: '^1.0.0',
>     rangeStrategy: 'replace',
>     currentVersion: '1.0.0',
>     newVersion: '1.0.7',
>   })
> ).toEqual('^1.0.7');
> ```
> 
> Do we really need upgrade to `^1.0.7` here? If I understand it right, `replace` strategy should return new ranges with new versions and shouldn't just shrink old ranges.

Agreed, it should return the existing range of `^1.0.0`.

File: lib/versioning/npm/index.spec.ts [hunk 3]
```diff
@@ -84,6 +84,16 @@ describe('semver.getNewValue()', () => {
       ['^1.2.3', '4.0.0', '^4.0.0'],
       ['^1.2.3', '4.5.6', '^4.0.0'],
       ['^1.0.0', '4.5.6', '^4.0.0'],
+
+      ['^0.2.3', '0.2.4', '^0.2.3'],
+      ['^2.3.0', '2.4.0', '^2.3.0'],
+      ['^2.3.4', '2.4.5', '^2.3.4'],
+      ['^0.0.1', '0.0.2', '^0.0.2'],
```
Comment (rarkins) [inline lib/versioning/npm/index.spec.ts:91]: I thought npm's semver treated `^0.0.1` as `>=0.0.1 <0.1.0`

Comment (zharinov) [reply lib/versioning/npm/index.spec.ts:91]: Good point, missed it

Comment (zharinov) [reply lib/versioning/npm/index.spec.ts:91]: https://github.com/npm/node-semver#caret-ranges-123-025-004
According to docs, it's `>=0.0.1 <0.0.2-0`

Comment (rarkins) [reply lib/versioning/npm/index.spec.ts:91]: Thanks, I was wrong. Leave as-is

**Review by rarkins — APPROVED** [review]

**Pull request merged**